extensions/mongodb/src/main/java/com/hazelcast/jet/mongodb/MongoDBSinkBuilder.java

```diff
@@ -19,173 +19,208 @@
-import java.util.ArrayList;
-import java.util.List;
 
-import static com.hazelcast.jet.impl.util.Util.checkSerializable;
 import static com.hazelcast.internal.util.Preconditions.checkNotNull;
+import static com.hazelcast.jet.impl.util.Util.checkNonNullAndSerializable;
+import static com.hazelcast.jet.impl.util.Util.checkSerializable;
+import static com.hazelcast.jet.retry.IntervalFunction.exponentialBackoffWithCap;
+import static com.mongodb.ReadPreference.primaryPreferred;
+import static com.mongodb.WriteConcern.MAJORITY;
+import static java.util.concurrent.TimeUnit.MINUTES;
 
 /**
- * See {@link MongoDBSinks#builder(String, SupplierEx)}
+ * See {@link MongoDBSinks#builder}
+ *
+ * @since 5.3
  *
  * @param <T> type of the items the sink will accept
  */
+@Beta
+@SuppressWarnings("UnusedReturnValue")
 public final class MongoDBSinkBuilder<T> {
 
-    private final String name;
-    private final SupplierEx<MongoClient> connectionSupplier;
+    /**
+     * Default transaction options used by the processors.
+     */
+    @SuppressWarnings("checkstyle:MagicNumber")
+    public static final TransactionOptions DEFAULT_TRANSACTION_OPTION = TransactionOptions
+            .builder()
+            .writeConcern(MAJORITY)
+            .maxCommitTime(10L, MINUTES)
+            .readPreference(primaryPreferred())
+            .build();
 
-    private FunctionEx<MongoClient, MongoDatabase> databaseFn;
-    private FunctionEx<MongoDatabase, MongoCollection<T>> collectionFn;
-    private ConsumerEx<MongoClient> destroyFn = ConsumerEx.noop();
+    /**
+     * Default retry strategy used by the processors.
+     */
+    @SuppressWarnings("checkstyle:MagicNumber")
+    public static final RetryStrategy DEFAULT_COMMIT_RETRY_STRATEGY = RetryStrategies
+            .custom()
+            .intervalFunction(exponentialBackoffWithCap(100, 2.0, 3000))
+            .maxAttempts(20)
+            .build();
+
+    private final String name;
+    private final WriteMongoParams<T> params = new WriteMongoParams<>();
 
-    private boolean ordered = true;
-    private boolean bypassValidation;
     private int preferredLocalParallelism = 2;
 
     /**
-     * See {@link MongoDBSinks#builder(String, SupplierEx)}
+     * See {@link MongoDBSinks#builder}
      */
     MongoDBSinkBuilder(
             @Nonnull String name,
-            @Nonnull SupplierEx<MongoClient> connectionSupplier
+            @Nonnull Class<T> documentClass,
+            @Nonnull SupplierEx<MongoClient> clientSupplier
     ) {
-        checkSerializable(connectionSupplier, "connectionSupplier");
-        this.name = name;
-        this.connectionSupplier = connectionSupplier;
+        this.name = checkNotNull(name, "sink name cannot be null");
+        params.setClientSupplier(checkNonNullAndSerializable(clientSupplier, "clientSupplier"));
+        params.setDocumentType(checkNotNull(documentClass, "document class cannot be null"));
+
+        if (Document.class.isAssignableFrom(documentClass)) {
```

Comment:
Document and BsonDocument needs to be elseif ?

Reply:
unfortunatelly yes - both have common supertype (Bson), but that type does not contain all necessary methods